cpp/src/arrow/flight/sql/driver.cc

```diff
@@ -0,0 +1,1877 @@
+                                   .Value(&info));
+      FlightInfoReader reader(client_, MakeCallOptions(CallContext::kFetch),
+                              std::move(info));
+      ADBC_RETURN_NOT_OK(reader.Init(error));
+      ADBC_ARROW_RETURN_NOT_OK(IO, error, IndexDbSchemas(&reader, &table_index, error));
+    }
+    if (depth == ADBC_OBJECT_DEPTH_ALL || depth >= ADBC_OBJECT_DEPTH_TABLES) {
+      std::unique_ptr<FlightInfo> info;
+      ADBC_ARROW_RETURN_NOT_OK(IO, error,
+                               client_
+                                   ->GetTables(call_options, /*catalog=*/nullptr,
+                                               db_schema ? &db_schema_filter : nullptr,
+                                               table_name ? &table_name_filter : nullptr,
+                                               /*include_schemas=*/true,
+                                               table_type ? &table_type_filter : nullptr)
+                                   .Value(&info));
+      FlightInfoReader reader(client_, MakeCallOptions(CallContext::kFetch),
+                              std::move(info));
+      ADBC_RETURN_NOT_OK(reader.Init(error));
+      ADBC_ARROW_RETURN_NOT_OK(IO, error,
+                               IndexTables(&reader, &table_index, &table_data, error));
+    }
+
+    std::unique_ptr<ArrayBuilder> raw_builder;
+    ADBC_ARROW_RETURN_NOT_OK(INTERNAL, error,
+                             MakeBuilder(kDbSchemaSchema).Value(&raw_builder));
+    StringBuilder catalog_name_builder;
+    ListBuilder catalog_db_schemas_builder(default_memory_pool(), std::move(raw_builder));
+    auto* catalog_db_schemas_items_builder =
+        checked_cast<StructBuilder*>(catalog_db_schemas_builder.value_builder());
+    auto* db_schema_name_builder =
+        checked_cast<StringBuilder*>(catalog_db_schemas_items_builder->field_builder(0));
+    auto* db_schema_tables_builder =
+        checked_cast<ListBuilder*>(catalog_db_schemas_items_builder->field_builder(1));
+    auto* db_schema_tables_items_builder =
+        checked_cast<StructBuilder*>(db_schema_tables_builder->value_builder());
+    auto* table_name_builder =
+        checked_cast<StringBuilder*>(db_schema_tables_items_builder->field_builder(0));
+    auto* table_type_builder =
+        checked_cast<StringBuilder*>(db_schema_tables_items_builder->field_builder(1));
+    auto* table_columns_builder =
+        checked_cast<ListBuilder*>(db_schema_tables_items_builder->field_builder(2));
+    auto* table_constraints_builder =
+        checked_cast<ListBuilder*>(db_schema_tables_items_builder->field_builder(3));
+    auto* table_columns_items_builder =
+        checked_cast<StructBuilder*>(table_columns_builder->value_builder());
+    auto* column_name_builder =
+        checked_cast<StringBuilder*>(table_columns_items_builder->field_builder(0));
+    auto* ordinal_position_builder =
+        checked_cast<Int32Builder*>(table_columns_items_builder->field_builder(1));
```

Comment:
You can perhaps presize some of these builders based on the table index and the table data length?
(not sure that matters, though)